sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,415 @@
+    pub fn position(&self, target_slot: Slot) -> Option<usize> {
+        let entries = self.as_entries_slice();
+        entries
+            .binary_search_by(|probe_entry| probe_entry.slot().cmp(&target_slot).reverse())
+            .ok()
+    }
+
+    /// Returns a `&[SlotHashEntry]` view into the underlying data.
+    ///
+    /// The constructor (in the safe path that called `parse_and_validate_data`)
+    /// or caller (if unsafe `new_unchecked` path) is responsible for ensuring
+    /// the slice is big enough and properly aligned.
+    #[inline(always)]
+    fn as_entries_slice(&self) -> &[SlotHashEntry] {
+        if self.len == 0 {
+            return &[];
+        }
+
+        debug_assert!(self.data.len() >= NUM_ENTRIES_SIZE + self.len * ENTRY_SIZE);
+
+        let entries_ptr =
+            unsafe { self.data.as_ptr().add(NUM_ENTRIES_SIZE) as *const SlotHashEntry };
+        unsafe { core::slice::from_raw_parts(entries_ptr, self.len) }
+    }
+
+    /// # Safety
+    /// Caller must ensure `index < self.len()`.
+    #[inline(always)]
+    pub unsafe fn get_entry_unchecked(&self, index: usize) -> &SlotHashEntry {
+        debug_assert!(index < self.len);
+        let offset = NUM_ENTRIES_SIZE + index * ENTRY_SIZE;
+        &*(self.data.as_ptr().add(offset) as *const SlotHashEntry)
+    }
+}
+
+impl<'a, T: Deref<Target = [u8]>> IntoIterator for &'a SlotHashes<T> {
+    type Item = &'a SlotHashEntry;
+    type IntoIter = core::slice::Iter<'a, SlotHashEntry>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.as_entries_slice().iter()
+    }
+}
+
+impl<'a> SlotHashes<Ref<'a, [u8]>> {
+    /// Creates a `SlotHashes` instance by safely borrowing data from an `AccountInfo`.
+    ///
+    /// This function verifies that:
+    /// - The account key matches the `SLOTHASHES_ID`
+    pub fn from_account_info(account_info: &'a AccountInfo) -> Result<Self, ProgramError> {
+        if account_info.key() != &SLOTHASHES_ID {
+            return Err(ProgramError::InvalidArgument);
+        }
+
+        let data_ref = account_info.try_borrow_data()?;
+
+        // Since the account key matches SLOTHASHES_ID, we can trust the runtime
+        // to have provided valid sysvar data
+        let num_entries = unsafe { read_entry_count_from_bytes_unchecked(&data_ref) };
+
+        Ok(unsafe { Self::new_unchecked(data_ref, num_entries) })
+    }
+}
+
+#[cfg(feature = "std")]
+impl SlotHashes<std::vec::Vec<u8>> {
+    /// Fetches the SlotHashes sysvar data directly via syscall. This copies
+    /// the full sysvar data (`MAX_SIZE` bytes).
+    pub fn fetch() -> Result<Self, ProgramError> {
+        let mut data = std::vec![0u8; MAX_SIZE];
+
+        // Use fetch_into to get the data and entry count
+        let num_entries = Self::fetch_into(&mut data, 0)?;
+
+        Ok(unsafe { Self::new_unchecked(data, num_entries) })
+    }
+}
+
+#[cfg(test)]
+#[path = "tests/slot_hashes_tests.rs"]
+mod slot_hashes_tests;
```

Comment:
I do not know if this is how tests are written in Pinocchio, but it seems a bit unusual.

Consider that if you put tests in a file called `slot_hashes/tests.rs` you get pretty much the same result as with this.  The tests are in a module of their own, and it is a submodule of `slot_hashes`.  But the advantage is that you do not need to break the conventions on the file name to module name mapping.

The fact that `sysvars/tests/slot_hashes_tests.rs` is a submodule of `sysvars/slot_hashes.rs` is quite unusual.
I didn't even know you could do this.

Reply:
There's no convention in pinocchio yet - this file was just growing too long, so it seems we're free to set the convention.

I agree that what I did here is much less common. I'll put tests into slot_hashes/tests.rs as you suggest.

But hey, now you know that this is possible, if unconventional:D